File: apps/web/core/api/observation-gateway.ts
import type {
  CorrectObservationRequest,
  HealthSuggestionDisposition,
  ImageAnalysisResult,
  Observation,
  ObservationListResult,
  ObservationPhotoPurpose,
  PlantJournalFrameListResult,
  RecordObservationRequest,
} from '@verdery/api-contracts';
import { IDEMPOTENCY_KEY_HEADER } from '@verdery/api-contracts';

import type { ApiClient } from './client';
import { csrfHeader } from './csrf';
import type { ApiResult } from './result';

/**
 * `ListPlantJournalFrames`' two optional parameters. `purpose` narrows the
 * sequence to one kind of shot, which is what makes consecutive frames
 * comparable; `limit` is a bound on the sequence, not a page size — there is
 * no cursor here, and asking for fewer frames means asking for a shorter
 * sequence.
 */
export interface PlantJournalFramesParams {
  readonly purpose?: ObservationPhotoPurpose | null;
  readonly limit?: number | null;
}

function journalFramesQuery(params: PlantJournalFramesParams | undefined): string {
  const search = new URLSearchParams();
  if (params?.purpose !== undefined && params.purpose !== null) {
    search.set('purpose', params.purpose);
  }
  if (params?.limit !== undefined && params.limit !== null) {
    search.set('limit', String(params.limit));
  }
  const query = search.toString();
  return query === '' ? '' : `?${query}`;
}

export interface ObservationGateway {
  record(
    gardenId: string,
    input: RecordObservationRequest,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Observation>>;
  listForGarden(gardenId: string, signal?: AbortSignal): Promise<ApiResult<ObservationListResult>>;
  listForPlant(
    gardenId: string,
    plantId: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<ObservationListResult>>;
  /** `ListPlantJournalFrames` (P11-MEDIA-01) — a plant's photographs oldest-first, for reading growth as a sequence. A read of what exists; nothing is rendered server-side. */
  listJournalFrames(
    gardenId: string,
    plantId: string,
    params?: PlantJournalFramesParams,
    signal?: AbortSignal,
  ): Promise<ApiResult<PlantJournalFrameListResult>>;
  correct(
    observationId: string,
    input: CorrectObservationRequest,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Observation>>;
  /** `SetHealthSuggestionDisposition` (P11-HEALTH-01) — no `If-Match`: `image_analysis_result` carries no revision, and a disposition may be reconsidered freely. */
  setHealthSuggestionDisposition(
    analysisResultId: string,
    disposition: HealthSuggestionDisposition,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<ImageAnalysisResult>>;
}

/**
 * Gateway for the observations-history endpoints.
 *
 * `record` never sends `photoMediaIds` from any current caller in
 * `features/observations`: this codebase has no upload flow yet, the same
 * gap `plant-gateway.ts`'s module doc comment explains, and
 * `RecordObservationRequest` already accepts a note and/or a condition
 * summary without a photo. `correct` carries no `If-Match`: `Observation` is
 * immutable and append-only, with no `revision` field to guard.
 *
 * Source: packages/api-contracts/openapi.yaml, tag `Observations`;
 * architecture/web-application-design.md, section "8. API Access".
 */
export function createObservationGateway(client: ApiClient): ObservationGateway {
  return {
    record(gardenId, input, idempotencyKey, signal) {
      return client.request<Observation>({
        method: 'POST',
        path: `/gardens/${gardenId}/observations`,
        body: input,
        headers: { [IDEMPOTENCY_KEY_HEADER]: idempotencyKey, ...csrfHeader() },
        ...(signal === undefined ? {} : { signal }),
      });
    },

    listForGarden(gardenId, signal) {
      return client.request<ObservationListResult>({
        method: 'GET',
        path: `/gardens/${gardenId}/observations`,
        ...(signal === undefined ? {} : { signal }),
      });
    },

    listForPlant(gardenId, plantId, signal) {
      return client.request<ObservationListResult>({
        method: 'GET',
        path: `/gardens/${gardenId}/plants/${plantId}/observations`,
        ...(signal === undefined ? {} : { signal }),
      });
    },

    listJournalFrames(gardenId, plantId, params, signal) {
      return client.request<PlantJournalFrameListResult>({
        method: 'GET',
        path: `/gardens/${gardenId}/plants/${plantId}/journal-frames${journalFramesQuery(params)}`,
        ...(signal === undefined ? {} : { signal }),
      });
    },

    correct(observationId, input, idempotencyKey, signal) {
      return client.request<Observation>({
        method: 'POST',
        path: `/observations/${observationId}/corrections`,
        body: input,
        headers: { [IDEMPOTENCY_KEY_HEADER]: idempotencyKey, ...csrfHeader() },
        ...(signal === undefined ? {} : { signal }),
      });
    },

    setHealthSuggestionDisposition(analysisResultId, disposition, idempotencyKey, signal) {
      return client.request<ImageAnalysisResult>({
        method: 'POST',
        path: `/observations/analysis-results/${analysisResultId}/disposition`,
        body: { disposition },
        headers: { [IDEMPOTENCY_KEY_HEADER]: idempotencyKey, ...csrfHeader() },
        ...(signal === undefined ? {} : { signal }),
      });
    },
  };
}

File: apps/ios/Sources/CoreNetworking/ObservationGateway.swift
import CoreDomain
import CoreObservability
import Foundation

/// The application's view of the observation history operations.
///
/// Features depend on this protocol, never on `URLSession` or a generated
/// client, so a feature test needs no network and no server — the same
/// reason `GardenGateway` exists.
///
/// Source: architecture/ios-application-design.md, section "9. Networking";
/// packages/api-contracts/openapi.yaml, tag `Observations`.
public protocol ObservationGateway: Sendable {
    /// Each attachment carries the shot purpose chosen for it
    /// (P11-MEDIA-01). Nothing is defaulted here: a purpose invented at this
    /// boundary would put the photograph into a comparison sequence it does
    /// not belong to.
    func recordObservation(
        gardenId: String,
        plantId: String?,
        gardenObjectId: String?,
        noteText: String?,
        conditionSummary: String?,
        observedAt: Date?,
        photos: [ObservationPhotoAttachment],
        idempotencyKey: String
    ) async throws -> GardenObservation

    func listObservationsForGarden(gardenId: String) async throws -> [GardenObservation]

    func listObservationsForPlant(gardenId: String, plantId: String) async throws -> [GardenObservation]

    /// `ListPlantJournalFrames` (P11-MEDIA-01): a plant's photographs in
    /// observed order, oldest first.
    ///
    /// `purpose` narrows the sequence to one kind of shot, which is what makes
    /// consecutive frames comparable; passing nil returns every photograph,
    /// including those carrying no label at all. `limit` is a bound on the
    /// sequence, not a page size — this operation has no cursor, and asking
    /// for fewer frames asks for a shorter sequence.
    func listPlantJournalFrames(
        gardenId: String,
        plantId: String,
        purpose: ObservationPhotoPurpose?,
        limit: Int?
    ) async throws -> [PlantJournalFrame]

    /// See `recordObservation`'s identical doc comment.
    func correctObservation(
        observationId: String,
        correctionKind: ObservationCorrectionKind,
        noteText: String?,
        conditionSummary: String?,
        photos: [ObservationPhotoAttachment],
        idempotencyKey: String
    ) async throws -> GardenObservation

    /// `SetHealthSuggestionDisposition` (P11-HEALTH-01): records the
    /// caller's disposition on an already-produced health suggestion. No
    /// `expectedRevision`/`If-Match` — `image_analysis_result` carries no
    /// revision, and a disposition may be reconsidered freely.
    func setHealthSuggestionDisposition(
        analysisResultId: String,
        disposition: HealthSuggestionDisposition,
        idempotencyKey: String
    ) async throws -> ImageAnalysisResult
}

/// URLSession-backed implementation of the observation history operations.
public struct URLSessionObservationGateway: ObservationGateway {
    private let transport: HTTPTransport

    public init(
        configuration: APIConfiguration,
        session: URLSession = .shared,
        correlationIdentifiers: any CorrelationIdentifierProvider =
            RandomCorrelationIdentifierProvider(),
        authTokenProvider: any AuthTokenProvider,
        appCheckTokenProvider: (any AppCheckTokenProvider)? = nil,
        log: any DiagnosticLog = NoOperationDiagnosticLog()
    ) {
        self.transport = HTTPTransport(
            configuration: configuration,
            session: session,
            correlationIdentifiers: correlationIdentifiers,
            authTokenProvider: authTokenProvider,
            appCheckTokenProvider: appCheckTokenProvider,
            log: log
        )
    }

    public func recordObservation(
        gardenId: String,
        plantId: String?,
        gardenObjectId: String?,
        noteText: String?,
        conditionSummary: String?,
        observedAt: Date?,
        photos: [ObservationPhotoAttachment],
        idempotencyKey: String
    ) async throws -> GardenObservation {
        let result: ObservationTransport = try await transport.send(
            method: "POST",
            operationPath: "gardens/\(gardenId)/observations",
            body: RecordObservationRequestTransport(
                plantId: plantId,
                gardenObjectId: gardenObjectId,
                noteText: noteText,
                conditionSummary: conditionSummary,
                observedAt: observedAt,
                photos: photos.map(ObservationPhotoAttachmentRequestTransport.init(attachment:))
            ),
            headers: [APIConfiguration.idempotencyKeyHeader: idempotencyKey],
            acceptedStatusCodes: [201]
        )
        return result.domainValue
    }

    public func listObservationsForGarden(gardenId: String) async throws -> [GardenObservation] {
        let result: ObservationListResultTransport = try await transport.get(
            operationPath: "gardens/\(gardenId)/observations",
            acceptedStatusCodes: [200]
        )
        return result.items.map(\.domainValue)
    }

    public func listObservationsForPlant(gardenId: String, plantId: String) async throws -> [GardenObservation] {
        let result: ObservationListResultTransport = try await transport.get(
            operationPath: "gardens/\(gardenId)/plants/\(plantId)/observations",
            acceptedStatusCodes: [200]
        )
        return result.items.map(\.domainValue)
    }

    public func listPlantJournalFrames(
        gardenId: String,
        plantId: String,
        purpose: ObservationPhotoPurpose?,
        limit: Int?
    ) async throws -> [PlantJournalFrame] {
        var queryItems: [String] = []
        if let purpose {
            queryItems.append("purpose=\(purpose.rawValue)")
        }
        if let limit {
            queryItems.append("limit=\(limit)")
        }

        var path = "gardens/\(gardenId)/plants/\(plantId)/journal-frames"
        if !queryItems.isEmpty {
            path += "?" + queryItems.joined(separator: "&")
        }

        let result: PlantJournalFrameListResultTransport = try await transport.get(
            operationPath: path,
            acceptedStatusCodes: [200]
        )
        return result.items.map(\.domainValue)
    }

    public func correctObservation(
        observationId: String,
        correctionKind: ObservationCorrectionKind,
        noteText: String?,
        conditionSummary: String?,
        photos: [ObservationPhotoAttachment],
        idempotencyKey: String
    ) async throws -> GardenObservation {
        let result: ObservationTransport = try await transport.send(
            method: "POST",
            operationPath: "observations/\(observationId)/corrections",
            body: CorrectObservationRequestTransport(
                correctionKind: correctionKind,
                noteText: noteText,
                conditionSummary: conditionSummary,
                photos: photos.map(ObservationPhotoAttachmentRequestTransport.init(attachment:))
            ),
            headers: [APIConfiguration.idempotencyKeyHeader: idempotencyKey],
            acceptedStatusCodes: [201]
        )
        return result.domainValue
    }

    public func setHealthSuggestionDisposition(
        analysisResultId: String,
        disposition: HealthSuggestionDisposition,
        idempotencyKey: String
    ) async throws -> ImageAnalysisResult {
        let result: ImageAnalysisResultTransport = try await transport.send(
            method: "POST",
            operationPath: "observations/analysis-results/\(analysisResultId)/disposition",
            body: SetHealthSuggestionDispositionRequestTransport(disposition: disposition),
            headers: [APIConfiguration.idempotencyKeyHeader: idempotencyKey],
            acceptedStatusCodes: [200]
        )
        return result.domainValue
    }
}
